### PR TITLE
Notification digest

### PR DESCRIPTION
**Pull request by pandurx**

### CODE REVIEW

**Comment by pandurx** [comment]
notification workflow

File: mod/cp_notifications/languages/fr.php [hunk 1]
```diff
@@ -308,16 +327,18 @@
...
+	'cp_notify:contactHelpDesk'=>'Si vous avez des questions, veuillez soumettre votre demande via le <a href="'.elgg_get_site_url().'mod/contactform/">formulaire Contactez-nous</a>.',
+    'cp_notify:visitTutorials'=>"Pour de plus amples renseignements sur GCconnex et ses fonctionnalités, consultez l'<a href='http://www.gcpedia.gc.ca/wiki/GCconnex_-_Aide_%C3%A0_l%27utilisateur/Modifier_mes_param%C3%A8tres_de_compte/Comment_puis-je_modifier_mes_param%C3%A8tres_de_notification%3F'>aide à l'utilisateur de GCconnex</a>.<br/>
 	                             Merci",
 
 	// cyu - new message in place for email content revamp
 	'cp_notify:french_follows' => "<i><sup>(Le fran&ccedil;ais suit)</sup></i>",
 	'cp_notify:readmore' => "<a href='%s'>Lire la suite</a>",
 
-    'cp_notify:email'=>'Courriel', // new
+	'cp_notify:personal_opportunities' => "Notify me when a new opportunity I have opted in for is created in the Career ConneXions Opportunities Platform",
```
Comment (Phanoix) [inline mod/cp_notifications/languages/fr.php]: Needs translation as well, just marking it since it seems a little buried here.

**Comment by Phanoix** [comment]
^ part 1 of my review.

File: mod/cp_notifications/actions/usersettings/save.php [hunk 2]
```diff
@@ -1,78 +1,105 @@
...
+		$result = $plugin->setUserSetting($k, $v, $user->guid);
+	}
+
+	/// create relationships between user and group that they have subscribed to
+	if (strpos($k,'cpn_site_mail_') !== false || strpos($k,'cpn_email_') !== false) {
+		
+		$group_guid = explode('_',$k);
+		if (strpos($k,'cpn_site_mail_') !== false) {
+		
+			if (strcmp($v,'set_notify_off') == 0)
```
Comment (pandurx) [inline mod/cp_notifications/actions/usersettings/save.php]: it returns default value (whatever you assigned it)

File: mod/cp_notifications/actions/usersettings/save.php [hunk 3]
```diff
@@ -1,78 +1,105 @@
...
+/// create the digest object for the user
+if (strcmp(elgg_get_plugin_user_setting('cpn_set_digest', $user->guid,'cp_notifications'),'set_digest_yes') == 0) {
+	if (!$user->cpn_newsletter) {
+
+		$new_digest = new ElggObject();
+		$new_digest->subtype = 'cp_digest';
+		$new_digest->owner_guid = $user->guid;
+		$new_digest->container_guid = $user->guid;
+		$new_digest->title = "Newsletter|{$user->email}";
+		$new_digest->access_id = ACCESS_PUBLIC;
```
Comment (pandurx) [inline mod/cp_notifications/actions/usersettings/save.php]: it needs to be public

File: mod/cp_notifications/actions/usersettings/save.php [hunk 4]
```diff
@@ -1,78 +1,105 @@
...
+		$group_guid = explode('_',$k);
+		if (strpos($k,'cpn_site_mail_') !== false) {
+		
+			if (strcmp($v,'set_notify_off') == 0)
+				add_entity_relationship($user->getGUID(), 'cp_subscribed_to_site_mail', $group_guid[sizeof($group_guid) - 1]);
+			else
+				remove_entity_relationship($user->getGUID(), 'cp_subscribed_to_site_mail', $group_guid[sizeof($group_guid) - 1]);
+		}
+	
+		if (strpos($k,'cpn_email_') !== false) {
```
Comment (pandurx) [inline mod/cp_notifications/actions/usersettings/save.php]: still working on code clean up for this mod

File: mod/cp_notifications/actions/usersettings/save.php (hunk 3 above)
Comment (Phanoix) [inline mod/cp_notifications/actions/usersettings/save.php]: ok, just checking

**Comment by Phanoix** [comment]
I'd suggest creating some unit tests for the lib functions sooner rather than later.

File: mod/cp_notifications/views/default/plugins/cp_notifications/usersettings.php [hunk 5]
```diff
@@ -1,410 +1,362 @@
...
-					} 
 
+/// PERSONAL SUBSCRIPTIONS (DISPLAYS ALL ITEMS THAT DO NOT BELONG IN GROUP NOTIFICATIONS)
 
-					// create a list of all the content in the group that you are subscribed to
-					for (var item in content_arr.output.text3)
-						$('#group-content-' + grp_guid).append("<div class='clearfix col-sm-12 list-break'>" + content_arr.output.text3[item] + "<div>");
+// cyu - patched issue with personal subscription that contains group content
+$query = "
+SELECT e.guid, e.subtype as entity_subtype, es.subtype, o.title, o.description
```
Comment (Phanoix) [inline mod/cp_notifications/views/default/plugins/cp_notifications/usersettings.php]: Have you tried running this query on a db like pre-prod? It doesn't look very large-database friendly...

File: mod/cp_notifications/lib/functions.php [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+	return true;
+}
+
+
+
+/**
+ * renders the correct subtype for the notification to display
+ * 
+ * @param string 			$subtype_name
+ * @param optional string 	$english
```
Comment (pandurx) [inline mod/cp_notifications/lib/functions.php]: Still cleaning in progress.

File: mod/cp_notifications/views/default/plugins/cp_notifications/usersettings.php (hunk 5 above)
Comment (pandurx) [inline mod/cp_notifications/views/default/plugins/cp_notifications/usersettings.php]: I have plans to make this Ajax. Will be part of code clean up (in progress).

**Comment by pandurx** [comment]
@KevinChagnon 
To view the admin notification troubleshoot tool:
1. Admin panel
2. in the Configure section of navigation 
3. Settings
4. Notification Email Messages Control Panel
5. Notifications Troubleshoot tab

**Pull request merged**